# Extraction of Digits in Python

In this notebook, we will learn how to extract digits from a number using basic arithmetic operations.

We will cover:
- Extracting digits using modulus and division
- Printing digits in reverse order
- Printing digits in correct order
- Counting total digits
- Finding sum of digits

This is a fundamental Data Structures & Algorithms (DSA) problem frequently asked in coding interviews.

In [2]:
def reverseDigits(num: int):
    n = num
    while n > 0:
        last_digit = n % 10
        print(last_digit, end="")
        n = n // 10


reverseDigits(4356)

6534

Step 1:

4356 % 10 = 6
Print → 6
4356 // 10 = 435

Step 2:

435 % 10 = 5
Print → 5
435 // 10 = 43

Step 3:

43 % 10 = 3
Print → 3
43 // 10 = 4

Step 4:

4 % 10 = 4
Print → 4
4 // 10 = 0

Loop stops because number becomes 0.

# Final Output
6534

Notice:
The digits are printed in reverse order because we always extract the last digit first.

⏱ Time Complexity

If a number has d digits:

Each loop removes one digit

Total iterations = number of digits

So,

Time Complexity = O(log n)
Space Complexity = O(1) (no extra storage used)

In [3]:
def countDigits(num: int):
    n = num
    count = 0
    
    while n > 0:
        n = n // 10
        count += 1
        
    return count

print(countDigits(4356))

4


# How Does This Work?

We use integer division (// 10) to remove one digit at a time.

Each time we remove a digit, we increase the counter.

When the number becomes 0, we stop.

The counter tells us how many digits the number had.

# Real-Life Example: Counting Currency Notes

Imagine you have ₹4356 written on a bill.

To count how many digits it has:

Remove the last digit → 435

Remove again → 43

Remove again → 4

Remove again → 0

We removed digits 4 times.

So, the number has 4 digits.

# Step-by-Step Execution for 4356

Initial:

n = 4356
count = 0

Step 1:

n = 4356 // 10 = 435
count = 1

Step 2:

n = 435 // 10 = 43
count = 2

Step 3:

n = 43 // 10 = 4
count = 3

Step 4:

n = 4 // 10 = 0
count = 4

Loop stops.

Final Answer:

4
⏱ Time and Space Complexity

If a number has d digits:

We divide the number d times.

Each division removes one digit.

So,

Time Complexity = O(log n)
Space Complexity = O(1)

In [4]:
import math

def countDigitsLog(num: int):
    if num == 0:
        return 1   # Special case: 0 has 1 digit
    return int(math.log10(num)) + 1

print(countDigitsLog(4356))

4


# Real-Life Intuition

Think of place values:

4356 can be written as:

4 × 10³ + 3 × 10² + 5 × 10¹ + 6 × 10⁰

The highest power here is 3.

So:

Number of digits = 3 + 1 = 4

That’s exactly what:

log₁₀(4356) ≈ 3.63
Floor(3.63) = 3
3 + 1 = 4

# Example Calculations

For n = 9
log₁₀(9) ≈ 0
Digits = 0 + 1 = 1

For n = 99
log₁₀(99) ≈ 1
Digits = 1 + 1 = 2

For n = 1000
log₁₀(1000) = 3
Digits = 3 + 1 = 4

⏱ Time & Space Complexity

Time Complexity = O(1)

Space Complexity = O(1)

This method is more efficient than the loop method (O(log n)).

## 🔁 Palindrome Number Check

A number is called a **Palindrome** if it reads the same forward and backward.

###  Examples

121 → Palindrome  
1331 → Palindrome  
4356 → Not a Palindrome  

---

## 💡 Approach (Without Converting to String)

We reverse the number mathematically and compare it with the original number.

In [ ]:
def isPalindrome(num: int):
    original = num
    reversed_num = 0
    
    while num > 0:
        last_digit = num % 10
        reversed_num = reversed_num * 10 + last_digit
        num = num // 10
        
    return original == reversed_num


print(isPalindrome(121))   
print(isPalindrome(4356))  

True
False


# How It Works (Step-by-Step Example: 121)

Initial:

original = 121
reversed_num = 0

Step 1:

last_digit = 121 % 10 = 1
reversed_num = 0 * 10 + 1 = 1
num = 121 // 10 = 12

Step 2:

last_digit = 12 % 10 = 2
reversed_num = 1 * 10 + 2 = 12
num = 12 // 10 = 1

Step 3:

last_digit = 1 % 10 = 1
reversed_num = 12 * 10 + 1 = 121
num = 1 // 10 = 0

Loop stops.

Now compare:

original == reversed_num
121 == 121 → True
# Real-Life Example

Think of a car number plate:

1221

If you see it in the mirror, it looks the same.

That’s exactly what we’re checking programmatically.

# Time & Space Complexity

If the number has d digits:

Time Complexity = O(log n)
Space Complexity = O(1)

We process each digit once.

# Edge Case

Negative numbers are NOT palindromes:

Example:
-121 → reversed becomes 121
But original is -121
So result is False.

## 💪 Armstrong Number (Narcissistic Number)

A number is called an **Armstrong Number** if:

The sum of each digit raised to the power of total number of digits  
is equal to the original number.

---

### 📌 Formula

If a number has `d` digits:

Armstrong condition:

(d₁^d + d₂^d + d₃^d + ... + dₙ^d) = original number

---

### 📍 Examples

153 → Armstrong  
1³ + 5³ + 3³  
= 1 + 125 + 27  
= 153 ✅  

9474 → Armstrong  
9⁴ + 4⁴ + 7⁴ + 4⁴  
= 6561 + 256 + 2401 + 256  
= 9474 ✅  

123 → Not Armstrong ❌ 

In [6]:
def isArmstrong(num: int):
    original = num
    digits = len(str(num))   # count number of digits
    total = 0
    
    while num > 0:
        last_digit = num % 10
        total += last_digit ** digits
        num = num // 10
        
    return total == original


print(isArmstrong(153))   
print(isArmstrong(9474))  
print(isArmstrong(123))   

True
True
False


# Step-by-Step Example (153)

Initial:
original = 153
digits = 3
total = 0

Step 1:
last_digit = 3
total = 0 + 3³ = 27

Step 2:
last_digit = 5
total = 27 + 5³ = 27 + 125 = 152

Step 3:
last_digit = 1
total = 152 + 1³ = 153

Now compare:

153 == 153 → True ✅

# Real-Life Intuition

Think of a "self-power number":

Each digit contributes its own power based on how many digits exist.

The number must reconstruct itself from its digits.

⏱ Time & Space Complexity

Let d be number of digits.

Time Complexity = O(log n)
Space Complexity = O(1)

We process each digit once.

⚠ Edge Cases

Single digit numbers (1–9) are always Armstrong.

Negative numbers are not considered Armstrong numbers.

